clients/src/test/java/org/apache/kafka/test/TestUtils.java

```diff
@@ -583,19 +583,6 @@ public static <T extends Throwable> void assertFutureThrows(
         assertEquals(expectedMessage, receivedException.getMessage());
     }
 
-    public static void assertFutureError(Future<?> future, Class<? extends Throwable> exceptionClass)
-        throws InterruptedException {
-        try {
-            future.get();
-            fail("Expected a " + exceptionClass.getSimpleName() + " exception, but got success.");
-        } catch (ExecutionException ee) {
-            Throwable cause = ee.getCause();
-            assertEquals(exceptionClass, cause.getClass(),
```

Comment:
The remaining method has a less useful message, it doesn't include the expected class. Perhaps we should address it as part of this change.

Reply:
The remaining method, `assertThrows`, appears as follows:

```java
private static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable, Object messageOrSupplier) {
  try {
    executable.execute();
  }
  catch (Throwable actualException) {
    if (expectedType.isInstance(actualException)) {
      return (T) actualException;
    }
    else {
      UnrecoverableExceptions.rethrowIfUnrecoverable(actualException);
      throw assertionFailure() //
        .message(messageOrSupplier) //
        .expected(expectedType) //
        .actual(actualException.getClass()) //
        .reason("Unexpected exception type thrown") //
        .cause(actualException) //
        .build();
    }
  }
  throw assertionFailure() //
    .message(messageOrSupplier) //
    .reason(format("Expected %s to be thrown, but nothing was thrown.", getCanonicalName(expectedType))) //
    .build();
}
```

This implementation seems to have the same effect as the `assertFutureError` method in [TestUtils.java](https://github.com/peterxcli/kafka/blob/b4be1785998d382843002ffd8685d67fe8d9c97d/clients/src/test/java/org/apache/kafka/test/TestUtils.java#L586-L597)

Would like to know your thoughts on this. Thanks!

Reply:
the replacement `assertFutureThrows` is using `assertThrows` and `assertInstanceOf`:
```java
    public static <T extends Throwable> T assertFutureThrows(Future<?> future, Class<T> exceptionCauseClass) {
        ExecutionException exception = assertThrows(ExecutionException.class, future::get);
        assertInstanceOf(exceptionCauseClass, exception.getCause(),
            "Unexpected exception cause " + exception.getCause());
        return exceptionCauseClass.cast(exception.getCause());
    }
```

I think the `expected class` is already included by the error message.

Reply:
I was referring to the `assertInstanceOf` part. I tested and this is what I see:

Before:
```org.opentest4j.AssertionFailedError: Expected a IllegalArgumentException exception, but got TimeoutException ==> 
Expected :class java.lang.IllegalArgumentException
Actual   :class org.apache.kafka.common.errors.TimeoutException
```

After:
```
org.opentest4j.AssertionFailedError: Unexpected exception cause org.apache.kafka.common.errors.TimeoutException: Timed out waiting for a node assignment. Call: createTopics ==> Unexpected type, 
Expected :class java.lang.IllegalArgumentException
Actual   :class org.apache.kafka.common.errors.TimeoutException
```

I think the new message is ok, but we should probably say something like "ExecutionException thrown by Future has unexpected cause ".

One additional detail, the new check is less strict - it allows the class or subclass while the old one only allowed the exact class. We should probably make the behavior configurable - if code expects an exact exception to be thrown, we should check for that.

Reply:
> One additional detail, the new check is less strict - it allows the class or subclass while the old one only allowed the exact class. We should probably make the behavior configurable - if code expects an exact exception to be thrown, we should check for that.

good point. Personally, I love the strict check. We could adjust `assertFutureThrows` to enforce this stricter behavior. While this might cause some existing tests to fail, I believe these failures should be relatively easy to address.

Reply:
That seems reasonable to me.